device/transport/mqtt/src/mqtt_twin_client.ts

```diff
@@ -0,0 +1,266 @@
+import * as dbg from 'debug';
+const debug = dbg('azure-iot-device-mqtt:MqttTwinClient');
+
+// $iothub/twin/PATCH/properties/reported/?$rid={request id}&$version={base version}
+
+/* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_009: [** The subscribed topic for `response` events shall be '$iothub/twin/res/#' **]** */
+const responseTopic = '$iothub/twin/res/#';
+
+/* Codes_SRS_NODE_DEVICE_MQTT_TWIN_RECEIVER_18_019: [** The subscribed topic for post events shall be $iothub/twin/PATCH/properties/desired/# **]** */
+const desiredPropertiesUpdatesTopic = '$iothub/twin/PATCH/properties/desired/#';
+
+interface TopicSubscription {
+  mqttClient: MqttBase;
+  topic: string;
+}
+
+/**
+ * @private
+ * @class        module:azure-iot-device-mqtt.MqttTwinReceiver
+ * @classdesc    Acts as a receiver for device-twin traffic
+ *
+ * @param {Object} config   configuration object
+ * @fires MqttTwinReceiver#subscribed   an MQTT topic has been successfully subscribed to
+ * @fires MqttTwinReceiver#error    an error has occured while subscribing to an MQTT topic
+ * @fires MqttTwinReceiver#response   a response message has been received from the service
+ * @fires MqttTwinReceiver#post a post message has been received from the service
+ * @throws {ReferenceError} If client parameter is falsy.
+ *
+ */
+export class MqttTwinClient extends EventEmitter {
+  static desiredPropertiesUpdateEvent: string = 'twinDesiredPropertiesUpdate';
+
+  private _mqtt: MqttBase;
+  private _pendingTwinRequests: { [key: string]: any } = {};
+  private _topicFsm: machina.BehavioralFsm;
+  private _responseTopic: TopicSubscription;
+  private _desiredPropertiesUpdatesTopic: TopicSubscription;
+
+  constructor(client: MqttBase) {
+    super();
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_18_001: [The `MqttTwinClient` constructor shall accept a `client` object.]*/
+    this._mqtt = client;
+
+    const messageHandler = this._onMqttMessage.bind(this);
+
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_001: [The `MqttTwinClient` constructor shall immediately subscribe to the `message` event of the `client` object.]*/
+    this._mqtt.on('message', messageHandler);
+
+    this._topicFsm = new machina.BehavioralFsm({
+      initialState: 'unsubscribed',
+      states: {
+        unsubscribed: {
+          _onEnter: function (client: TopicSubscription, err: Error, callback: (err?: Error) => void): void {
+            if (callback) {
+              callback(err);
+            }
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            this.transition(client, 'subscribing', callback);
+          },
+          unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            // not entirely sure about that. if subscription are restored because cleanSession is false, it means technically a user may want to unsubscribe
+            // even though subscribe hasn't been called yet.
```

Comment:
Indeed.  I am pondering this issue in another issue.

Reply:
There is also lore and history here around this issue.  IIRC, the service team wanted us to _not_ unsubscribe.  If we do unsubscribe here, then the service client class that updates the desired properties will fail of the device is offline.  Keeping the topic subscribed allows the service API call to succeed.  (no kidding)  Unless they changed this behavior while I was looking the other way.

---
In reply to: [170587012](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170587012) [](ancestors = 170587012)

Reply:
Wait, my comment actually applies to the "unsubscribing" block below, and not here, but it still stands.

---
In reply to: [170698604](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170698604) [](ancestors = 170698604,170587012)

Reply:
Well this depends on who is calling this level.  The service does indeed NOT want us to unsubscribe when when disconnect/close.  In general this could lead to a pattern when something "above" us here re-connects, it might actually want to unsubscribe to a topic without having done a subscribe first.  Now whether this will happen since this is somewhat of a private interface ....

---
In reply to: [170699009](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170699009) [](ancestors = 170699009,170698604,170587012)

Reply:
I'll get to the bottom of this, thanks for pointing it out. may or may not be solved with this PR though.

---
In reply to: [170703095](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170703095) [](ancestors = 170703095,170699009,170698604,170587012)

Reply:
apparently we should unsubscribe when we disconnect.

---
In reply to: [171442744](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r171442744) [](ancestors = 171442744,170703095,170699009,170698604,170587012)